lib/pbench/agent/tool_meister.py

```diff
@@ -110,20 +127,113 @@ def __init__(
         assert (
             pbench_install_dir is not None
         ), "Logic bomb!  no installation directory provided!"
+        assert (
+            pbench_install_dir.is_dir()
+        ), f"pbench installation directory does not exist: {pbench_install_dir}"
         self.pbench_install_dir = pbench_install_dir
+        assert tool_dir is not None, "Logic bomb!  no tool directory provided!"
+        assert tool_dir.is_dir(), f"tool directory does not exist: {tool_dir}"
         self.tool_dir = tool_dir
-        self.start_process = None
-        self.stop_process = None
 
-    def _check_no_processes(self):
-        if self.start_process is not None:
-            raise ToolException(
-                f"Tool({self.name}) has an unexpected start process running"
+    def install(self):
+        raise NotImplementedError(
+            f"{self.__class__.__name__} does not implement the install method"
+        )
+
+    def start(self):
+        raise NotImplementedError(
+            f"{self.__class____name__} does not implement the start method"
+        )
+
+    def stop(self):
+        raise NotImplementedError(
+            f"{self.__class____name__} does not implement the stop method"
+        )
+
+    def wait(self):
+        raise NotImplementedError(
+            f"{self.__class____name__} does not implement the wait method"
+        )
+
+    def _wait_for_process(self, process, ctx_name=None):
+        """Generic method to wait for a given process to stop after 30
+        seconds, emitting a message every 5 seconds in between.
+
+        Returns the process return code on success, or None if the process
+        failed to exit.
+        """
+        count = 6
+        sts = None
+        while count > 0 and sts is None:
+            try:
+                sts = process.wait(timeout=5)
+            except subprocess.TimeoutExpired:
+                count -= 1
+                if count > 0:
+                    self.logger.debug(
+                        "%s tool %s%s has not stopped after %d seconds",
+                        self._tool_type,
+                        self.name,
+                        f" {ctx_name}" if ctx_name else "",
+                        5 * (6 - count),
+                    )
+        return sts
```

Comment:
I think I would have written this something like:
```
        intervals = 6
        timeout = 5
        for count in range(0, timeout * intervals, timeout):
            try:
                return process.wait(timeout=timeout)
            except subprocess.TimeoutExpired:
                self.logger.debug(
                        "%s tool %s%s has not stopped after %d seconds",
                        self._tool_type,
                        self.name,
                        f" {ctx_name}" if ctx_name else "",
                    count + timeout
                )
        return None
```

Reply:
Allow me to address this in another commit coming.